ietf/meeting/helpers.py

```diff
@@ -243,39 +244,401 @@ def preprocess_assignments_for_agenda(assignments_queryset, meeting, extra_prefe
 
     return assignments
 
-def is_regular_agenda_filter_group(group):
-    """Should this group appear in the 'regular' agenda filter button lists?"""
-    return group.type_id in ('wg', 'rg', 'ag', 'rag', 'iab', 'program')
 
-def tag_assignments_with_filter_keywords(assignments):
-    """Add keywords for agenda filtering
-    
-    Keywords are all lower case.
+class AgendaKeywordTool:
+    """Base class for agenda keyword-related organizers
+
+    The purpose of this class is to hold utility methods and data needed by keyword generation
+    helper classes. It ensures consistency of, e.g., definitions of when to use legacy keywords or what
+    timeslot types should be used to define filters.
     """
-    for a in assignments:
-        a.filter_keywords = {a.timeslot.type.slug.lower()}
-        a.filter_keywords.update(filter_keywords_for_session(a.session))
-        a.filter_keywords = sorted(list(a.filter_keywords))
-
-def filter_keywords_for_session(session):
-    keywords = {session.type.slug.lower()}
-    group = getattr(session, 'historic_group', session.group)
-    if group is not None:
-        if group.state_id == 'bof':
-            keywords.add('bof')
-        keywords.add(group.acronym.lower())
-        area = getattr(group, 'historic_parent', group.parent)
-
-        # Only sessions belonging to "regular" groups should respond to the
-        # parent group filter keyword (often the 'area'). This must match
-        # the test used by the agenda() view to decide whether a group
-        # gets an area or non-area filter button.
-        if is_regular_agenda_filter_group(group) and area is not None:
-            keywords.add(area.acronym.lower())
-    office_hours_match = re.match(r'^ *\w+(?: +\w+)* +office hours *$', session.name, re.IGNORECASE)
-    if office_hours_match is not None:
-        keywords.update(['officehours', session.name.lower().replace(' ', '')])
-    return sorted(list(keywords))
+    filterable_timeslot_slugs = ['officehours']
+    filterable_timeslot_types = TimeSlotTypeName.objects.filter(slug__in=filterable_timeslot_slugs)
+
+    def __init__(self, *, assignments=None, sessions=None):
+        # n.b., single star argument means only keyword parameters are allowed when calling constructor
+        if assignments is not None and sessions is None:
+            self.assignments = assignments
+            self.sessions = [a.session for a in self.assignments if a.session]
```

Comment:
I think it will bomb out with a `ValueError` or `KeyError` if it gets a scalar for either assignments or sessions. That seems pretty close to the right thing to do. It should point at the right place in the traceback, so I don't know that catching the error just to improve its error message adds a whole lot of value.

I can add type hints to help the linter understand out that these need to be iterables.

Reply:
Whatever you think is best - you're much more the expert than I.